test/prefer-default-parameters.js

```diff
@@ -0,0 +1,156 @@
+import test from 'ava';
+import avaRuleTester from 'eslint-ava-rule-tester';
+import {outdent} from 'outdent';
+import rule from '../rules/prefer-default-parameters';
+
+const ruleTester = avaRuleTester(test, {
+	env: {
+		es6: true
+	}
+});
+
+const error = {
+	ruleId: 'prefer-default-parameters',
+	messageId: 'preferDefaultParameters'
+};
+
+ruleTester.run('prefer-default-parameters', rule, {
+	valid: [
+		'function abc(foo = { bar: 123 }) { }',
+		'function abc({ bar } = { bar: 123 }) { }',
+		'function abc({ bar = 123 } = { bar }) { }',
+		'function abc(foo = fooDefault) { }',
+		'function abc(foo = {}) { }',
+		'function abc(foo = \'bar\') { }',
+		'function abc({ bar = 123 } = {}) { }',
+		'function abc(foo) { foo = foo || bar(); }',
+		'function abc(foo) { foo = foo || {bar} }',
+		'function abc(foo) { const {bar} = foo || 123 }',
+		'function abc(foo, bar) { bar = foo || \'bar\' }',
+		'function abc(foo) { foo = foo || 1 && 2 || 3 }',
+		'function abc(foo) { foo = !foo || \'bar\' }',
+		'function abc(foo) { foo = (foo && bar) || baz }',
+		'function abc(foo = 123) { foo = foo || \'bar\' }',
+		'function abc() { let foo = 123; foo = foo || \'bar\' }',
+		'function abc() { let foo = 123; const bar = foo || \'bar\' }',
+		'const abc = (foo = \'bar\') => { };',
+		'const abc = (foo, bar) => { bar = foo || \'bar\' };',
+		'foo = foo || \'bar\';',
+		'const bar = foo || \'bar\';',
+		outdent`
+			function abc(foo) {
+				function def(bar) {
+					foo = foo || 'bar';
+				}
+			}
+		`
+	],
+	invalid: [
+		{
+			code: 'function abc(foo) { foo = foo || \'bar\'; }',
+			output: 'function abc(foo = \'bar\') { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo) { foo = foo || \'bar\';}',
+			output: 'function abc(foo = \'bar\') { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo) { foo = foo || 123; }',
+			output: 'function abc(foo = 123) { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo) { foo = foo || true; }',
+			output: 'function abc(foo = true) { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo, bar) { foo = foo || \'bar\'; baz(); }',
+			output: 'function abc(foo = \'bar\', bar) { baz(); }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo) { const bar = foo || \'bar\'; }',
+			output: 'function abc(bar = \'bar\') { }',
+			errors: [error]
+		},
+		{
+			code: 'function abc(foo) { let bar = foo || \'bar\'; }',
+			output: 'function abc(bar = \'bar\') { }',
```

Comment:
Not sure about this `function abc(foo = \'bar\') { const bar = foo;}`, this should easier, but you already did better. Any chance this would break something?

Reply:
Yeah, if some code relies on the presence of the `foo` parameter it might berak. I'm going to use the suggestions API for this.

Reply:
We can check if `foo` is used. If not used, in the way you already did, otherwise fix to the code I wrote.

WDYT?